smarts/env/wrappers/metrics.py

```diff
@@ -0,0 +1,273 @@
+
+    counts: Counts
+    costs: Costs
+
+
+@dataclass
+class Data:
+    """Stores an agent's performance record and cost functions."""
+
+    record: Record
+    cost_funcs: CostFuncs
+
+
+class Metrics(gym.Wrapper):
+    """Metrics class wraps an underlying _Metrics class. The underlying
+    _Metrics class computes agents' performance metrics in a SMARTS
+    environment. Whereas, this Metrics class is a basic gym.Wrapper class
+    which prevents external users from accessing or modifying attributes
+    beginning with an underscore, to ensure security of the metrics computed.
+
+    Args:
+        env (gym.Env): A gym.Env to be wrapped.
+
+    Raises:
+        AttributeError: Upon accessing an attribute beginning with an underscore.
+
+    Returns:
+        gym.Env: A wrapped gym.Env which computes agents' performance metrics.
+    """
+
+    def __init__(self, env: gym.Env):
+        env = _Metrics(env)
+        super().__init__(env)
+
+
+class _Metrics(gym.Wrapper):
+    """Computes agents' performance metrics in a SMARTS environment."""
+
+    def __init__(self, env: gym.Env):
+        super().__init__(env)
+        _check_env(env)
+        self._cur_scen: str
+        self._cur_agents: Set[str]
+        self._steps: Dict[str, int]
+        self._done_check: Set[str]
+        self._records = {}
+
+    def step(self, action: Dict[str, Any]):
+        """Steps the environment by one step."""
+        obs, rewards, dones, info = super().step(action)
+
+        # Only count steps in which an ego agent is present.
+        if len(obs) == 0:
+            return obs, rewards, dones, info
+
+        for agent_name, agent_obs in obs.items():
+            self._steps[agent_name] += 1
+
+            # Compute all cost functions.
+            # fmt: off
+            costs = Costs()
+            for field in fields(self._records[self._cur_scen][agent_name].cost_funcs):
+                cost_func = getattr(self._records[self._cur_scen][agent_name].cost_funcs, field.name)
+                new_costs = cost_func(agent_obs)
+                costs = _add_dataclass(new_costs, costs)
+            # fmt: on
+
+            # Update stored costs.
+            self._records[self._cur_scen][agent_name].record.costs = costs
+
+            if dones[agent_name]:
+                self._done_check.add(agent_name)
+                steps_adjusted = 0
+                goals = 0
+                crashes = 0
+                reason = termination.reason(obs=agent_obs)
+                if reason == termination.Reason.Goal:
+                    steps_adjusted = self._steps[agent_name]
+                    goals = 1
+                elif reason == termination.Reason.Crash:
+                    steps_adjusted = _MAX_STEPS
+                    crashes = 1
+                else:
+                    raise Exception(
+                        f"Unsupported agent done reason. Events: {agent_obs.events}."
+                    )
```

Comment:
I think we should raise a more specific error than `Exception` since `Exception` makes the error hard to isolate for purposes of debugging and `try-catch`.

Reply:
A custom `MetricsError` class has been added.